src/Cryptol/TypeCheck/Solve.hs

```diff
@@ -287,15 +289,16 @@ proveImplication lnam as ps gs =
 
 
 proveImplicationIO :: Solver
+                   -> Bool     -- ^ Whether to remove duplicate goals in errors
```

Comment:
I wonder if really need the extra argument, isn't it reasonable to always remove duplicate goals?

Reply:
`Goal` has a weird `Eq` and `Ord` instance where it only compares the actual `Prop` and not where it came from, so I thought it might still be useful to report if you needed to satisfy the same `Prop` but in two different places. In the case of foreign declarations the source is always the same (it just says it's coming from that foreign declaration) so it's safe to deduplicate it by comparing `Prop`s.

I could have also deduplicated by comparing all the fields, but that would have required adding an `Eq` instance for `ConstraintSource` and then for `Expr`, which seemed like a lot of changes and I wasn't sure if `Expr` lacked an `Eq` instance for a reason.